Enable to use disinct NotFoundError for each model

muddydixon opened the pull request:
In current version every `NotFoundError` is same as `Errors.NotFoundError`, thus we cannot discriminate the NotFoundError occurred from.

The p-r enables this problem by setting in static properties bellow:

```coffee-script
Foo = bookshelf.Model.extend {
  tableName: "foo"
}, {
  NotFoundError: createError(bookshelf.Model.NotFoundError, "FooNotFoundError")
}

Bar = bookshelf.Model.extend {
  tableName: "bar"
}, {
  NotFoundError: createError(bookshelf.Model.NotFoundError, "BarNotFoundError")
}

Baz = bookshelf.Model.extend {
  tableName: "baz"
}

console.log Bar.NotFoundError is Foo.NotFoundError # => false

Foo.where(name: "foo").fetch(require: true)
.catch((err)->
  console.log err # => { [FooNotFoundError: EmptyResponse] message: 'EmptyResponse' }
  console.log err instanceof bookshelf.Model.NotFoundError # => true
  console.log err instanceof Foo.NotFoundError # => true
  console.log err instanceof Bar.NotFoundError # => false
  console.log err instanceof Baz.NotFoundError # => true use default because not override
)
```

bendrucker commented:
Thanks @muddydixon! I really like this as a way of addressing #674. I'm really not keen on doing any sort of dynamic magic with `extend` since it takes away the option to use the `extends` keyword in languages that support it. This is a nice middle ground.

If you add this consistently across all the custom errors I'll get it merged right away.

muddydixon commented:
I see, i will be going to try solutions. But please merge this pr if you can.

I think it better that replace `simple-extend` module.

bendrucker commented:
Seems like we're not on the same page here. 

If you add `NoRowsDeletedError` and `NoRowsUpdatedError` to the `Model` and add the same behavior for `Collection.EmptyError`, this is ready to merge. But it has to be consistent.

muddydixon commented:
Oops, I misunderstand your rep. 
Ok, i try it!

bendrucker commented:
Yup, figured it was just a language barrier. Thanks again!

muddydixon commented:
Done! check plz!

xaka commented:
@bendrucker This PR doesn't actually address #674 as it still references `BookshelfModel.NotFoundError` implicitly. To properly address that issue we'd have to create a separate `NotFoundError` class for each new model users define using `extend`. Unless I'm missing something here.

tgriesser commented:
I think the point was that you could supply your own if you wanted. Which I forget who i was talking with about the other day, possibly @joepie91 who was looking for something similar. 

I guess that seems like a reasonable enough request to have an error class per-constructor

joepie91 commented:
For clarification - the thing I was describing was to automatically create an "error subclass" for each model, so that you could handle `NotFoundError`s by looking for eg. `User.NotFoundError` or `Post.NotFoundError`.

muddydixon commented:
@xaka Yes `BookshelfModel.NotFoundError` is used implicitly because of backward compatibility.
If `SomeModel.NotFoundError` is used without overriding `NotFoundError` in static Properties, exception is not occurred.
But I agree your opinion, I want to implement distinct `NotFoundError` for each model that is extended from `Errors.NotFoundError` too.

Its not ideal, anyway my p-r solves addressing problem to some extent,

xaka commented:
@muddydixon I do believe this PR is a good start and it leads us towards what we all want to see at the end, which is automatic error sub classes per model. After the merge we can go to #674 to continue conversation regarding the best way to get it done.

muddydixon commented:
@xaka I sent new commit in that per-model Errors were set in default.
Please review it.

xaka commented:
I don't think @bendrucker is going to like it :) as he was against using `simple-extend`, but with a little input/suggestion I believe we can finally nail it down.

tgriesser commented:
It's fine, the next version of bookshelf is going to decorate constructors while maintaining support for ES6 `extends`, and drop simple-extend for a custom extend, so there's no problem in adding in a little bit of magic via `.extended`

xaka commented on bookshelf.js:
```diff
@@ -21,6 +21,7 @@ Bookshelf.initialize = function(knex) {
   var _          = require('lodash');
   var inherits   = require('inherits');
   var semver     = require('semver');
+  var createError = require('create-error');
```
You probably want to keep it align with other lines so it looks consistent

tgriesser commented:
If you drop the "self" in favor of "this" and not use `_.capitalize` since we are currently supporting back to 2.x version of lodash, and add a test I'll go ahead and merge.

xaka commented on bookshelf.js:
```diff
@@ -149,10 +150,22 @@ Bookshelf.initialize = function(knex) {
+  };
+  Collection.extend = function(protoProps, staticProps){
+    this.EmptyError = createError(BookshelfCollection.EmptyError,
+      _.capitalize(protoProps.tableName) + "EmptyError");
```
Looks like the bookshelf is using single quotes so you probably want to replace double quotes here

xaka commented:
@muddydixon Let me know if you need help with adding tests so we could keep wheels rolling. I need this one so bad :)

muddydixon commented:
I'm sorry for late.

In previous commit, Model own properties are overridden in each time to `extend`.

For example,

```
var Hoge = bookshelf.Model.extend({
  tableName: "bookshelf"
});
// Model.NotFoundError => "HogeNotFoundError"
var Fuga = bookshelf.Model.extend({
  tableName: "bookshelf"
});
// Model.NotFoundError => "FugaNotFoundError"
// thus

new Hoge.NotFoundError() // => "FugaNotFoundError" awesome!!!
```

I fixed this!

And added test.
